### PR TITLE
State sessions

### PR DESCRIPTION
**Pull request by richard-to**
# Before 
before

# After
after

-----

Here's an initial implementation. Mainly want to make sure I'm on a decent track here before continuing on since there's a few differences from what we've discussed:

1) I tried to implement a me.configure_state option, but was having trouble getting it to load the configuration before the Flask app started. If we won't need the Flask Secret Key, then not a big deal. Here'a rough implementation here: https://github.com/richard-to/mesop/commit/d3b5b159acc74f8d1ab80b48af0cf2b9fd29401b. So here, ended up using the environment variable approach for now.

2) We discussed using the hash of the state as the key. During implementation I realized we don't really need to have a hash of the state since all we're checking is if we have the state cached. So basically any key will do. Especially for the Memory backend, it didn't make sense to serialize the state data just to generate a hash which wouldn't be guaranteed to be unique anyway especially for simple state objects. So I ended up doing two things. One was generating a temporary token to use as as the key. This key will get passed by in the UI response. The drawback of this is that we generate a new token every time. So I also implemented the Flask session token as the primary option. The temporary token option is used when no secret key is set (though we could mitigate this if we figure out a way to generate reasonable secret key for the user -- wasn't quite sure how to get the main entry point file for hashing)

3) One thing I couldn't figure how to do was to handle the retry on client side when the cache key is not found

----

Ref: https://github.com/google/mesop/issues/62

### CODE REVIEW

**Comment by wwwillchen** [comment]
> Here's an initial implementation. Mainly want to make sure I'm on a decent track here before continuing on since there's a few differences from what we've discussed:
> 
> 1. I tried to implement a me.configure_state option, but was having trouble getting it to load the configuration before the Flask app started. If we won't need the Flask Secret Key, then not a big deal. Here'a rough implementation here: [richard-to@d3b5b15](https://github.com/richard-to/mesop/commit/d3b5b159acc74f8d1ab80b48af0cf2b9fd29401b). So here, ended up using the environment variable approach for now.

sounds OK

> 2. We discussed using the hash of the state as the key. During implementation I realized we don't really need to have a hash of the state since all we're checking is if we have the state cached. So basically any key will do. Especially for the Memory backend, it didn't make sense to serialize the state data just to generate a hash which wouldn't be guaranteed to be unique anyway especially for simple state objects. So I ended up doing two things. One was generating a temporary token to use as as the key. This key will get passed by in the UI response. The drawback of this is that we generate a new token every time. So I also implemented the Flask session token as the primary option. The temporary token option is used when no secret key is set (though we could mitigate this if we figure out a way to generate reasonable secret key for the user -- wasn't quite sure how to get the main entry point file for hashing)

Good point, the hashing is only needed if we want to share state across users/sessions, but that seems a bit sketchy.

What's wrong with generating a new token each time?

I thought more about the main entry point file for hashing, and I'm not sure how it would work in Colab use case, since there's not really a regular Python module file. Otherwise, you could use [`inspect.getsource`](https://docs.python.org/3/library/inspect.html#inspect.getsource)

> 3. One thing I couldn't figure how to do was to handle the retry on client side when the cache key is not found

File: mesop/server/server.py [hunk 1]
```diff
@@ -125,7 +128,14 @@ def generate_data(ui_request: pb.UiRequest) -> Generator[str, None, None]:
       elif ui_request.HasField("user_event"):
         event = ui_request.user_event
         runtime().context().set_viewport_size(event.viewport_size)
-        runtime().context().update_state(event.states)
+
+        if event.states.states:
+          runtime().context().update_state(event.states)
+        else:
```
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/server/server.py:134]: Ah good point about that scenario for the memory case. It wouldn't recover correctly, so we'd still need to generate a some kind of unique token along with the uuid. Which in that case, maybe we can just rely on generating the random token each time.

--------

I think the reason for putting the UUID in the Flask Session is mainly to avoid tampering. If the UUID were passed in with the request, a user could use it from another client. However I guess it would still be possible to do this with the session cookie as well. But I think it would be harder for someone to get the session cookie than getting access to the UUID from the response?

But I think that's moot since we do need to generate a unique token each time or calculate the state hash.

File: mesop/web/src/services/channel.ts [hunk 2]
```diff
@@ -196,6 +202,13 @@ export class Channel {
             break;
           }
           case UiResponse.TypeCase.ERROR:
+            if (
+              uiResponse.getError()?.getException() ===
+              'No state hash found in state session backend.'
+            ) {
+              // How to retry?
+              // this.init(this.initParams, request);
+            }
```
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/web/src/services/channel.ts:211]: Yeah, it didn't work for me. The retry would get "cancelled" in the network tab. So it would never complete. Not sure why. But I didn't dig too deep yet.

**Comment by richard-to** [comment]
> What's wrong with generating a new token each time?

Not too much. Just seemed like a waste to keep recomputing tokens on every request. But after your other comment, seems like we'd need to generate a new token each time or calculate the state hash to ensure we're able to recover correctly for the in memory case.

> I thought more about the main entry point file for hashing, and I'm not sure how it would work in Colab use case, since there's not really a regular Python module file. Otherwise, you could use [`inspect.getsource`](https://docs.python.org/3/library/inspect.html#inspect.getsource)

I guess I'm not sure how to get the module file that is used as the entry point. I think I can get it if I could access the path flag in cli.py. But I then I think for the pip version, the path flag is not used. Instead it's the first command line argument. And as you mentioned the colab case.

But I think it seems ok if the secret key is not set. Then the fallback to generating the random tokens seems necessary anyway. So wouldn't be much different if the user set the secret key. In this later case it would be the UUID+RandomToken. UUID would be passed in the session. And RandomToken as part of the response. Main benefit there would be lower risk of clashing tokens. Though I don't know what the likelihood of collisions when using the secrets library.

File: mesop/server/server.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/server/server.py:134]: So I think what I'll do is just remove secret key part and rely solely on regenerating the token after each request. And I think we can prepend a timestamp to the token to ensure a bit more uniqueness.

[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/server/server.py:134]: Ok. Removed the Flask session stuff for now.

File: mesop/server/config.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+from pydantic import BaseModel
+
+load_dotenv()
+
+
+class Config(BaseModel):
+  """Mesop app configuration."""
+
+  secret_key: str = ""
+  state_session_backend: Literal["memory", "null"] = "memory"
```
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/server/config.py]: Yes, the issue I ran into here is that if I did None, the user would need to specify an environment variable, otherwise if not set the default would be none. And if I set a default if the environment variable was not set, then None would not be an option.

I guess in this case I could make it so empty string gets converted to None.

In the end, I renamed "null" to "none". But still need to explicitly specify "none"

File: mesop/web/src/services/channel.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (wwwillchen) [reply mesop/web/src/services/channel.ts:213]: I think this could be because if you do `this.init` here, the original EventSource hasn't closed yet, and in L104 https://github.com/google/mesop/blob/c7aefdc62eb040d6d2f0a93ba253d0e46d2f6ce6/mesop/web/src/services/channel.ts#L104 it'll close the EventSource *after* you've done the retry with this.init (which is why I think you saw the retry get cancelled).

Instead if you put this in this.queuedEvents (probably the front of the queue), similar to https://github.com/google/mesop/blob/c7aefdc62eb040d6d2f0a93ba253d0e46d2f6ce6/mesop/web/src/services/channel.ts#L231, then it should work

[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/web/src/services/channel.ts]: Nice adding it to the queue worked.  I think the only thing left here is handling the case where the retry keeps failing, so we probably want to give up after a couple failures.

[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/web/src/services/channel.ts]: Actually, I think it's OK. There should never be an infinite loop since the retry includes the state. In the worst case the memory cache is not working at all and every request takes two tries.

**Comment by richard-to** [comment]
I'm running into this type checking issue that I'm not quite sure how to workaround. I looked online but haven't found an explanation that makes sense to me.

```
  /home/runner/work/mesop/mesop/mesop/server/config.py:22:27 - error: Argument of type "str" cannot be assigned to parameter "state_session_backend" of type "Literal['memory', 'none']" in function "__init__"
    Type "str" cannot be assigned to type "Literal['memory', 'none']"
      "str" cannot be assigned to type "Literal['memory']"
      "str" cannot be assigned to type "Literal['none']" (reportGeneralTypeIssues)
```

File: mesop/server/state_session.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+
+class MemoryStateSessionBackend(StateSessionBackend):
+  """Stores state in memory.
+
+  The main limitation of this backend is that deloyments with multiple processes /
+  replicas / instances will cause a lot of cache misses if there is no session affinity.
+
+  This is because memory is local to each Mesop server process.
+
```
Comment (wwwillchen) [inline mesop/server/state_session.py:45]: I'm wondering if we should check how much memory is left. Otherwise, if there's a bunch of sessions with large states (e.g. 100MB dataframe), will the server hit out of memory errors (OOM)?

Could use a library like:
https://github.com/giampaolo/psutil?tab=readme-ov-file#memory

File: mesop/server/config.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  state_session_backend: Literal["memory", "none"] = "memory"
+
+  @property
+  def state_session_enabled(self):
+    return self.state_session_backend != "none"
+
+
+def CreateConfigFromEnv() -> Config:
+  return Config(
+    state_session_backend=os.getenv("MESOP_STATE_SESSION_BACKEND", "memory"),
```
Comment (wwwillchen) [inline mesop/server/config.py:22]: I think we should make this opt-in initially - I'm a little concerned we will hit OOMs (see other comment), and then later make this the default

File: mesop/server/state_session.py (hunk 4 above)
Comment (wwwillchen) [reply mesop/server/state_session.py:45]: [By default, Cloud run has 512MB](https://cloud.google.com/run/docs/configuring/services/memory-limits#setting) so I think we need to be fairly conservative with how much memory we allocate for this

Comment (wwwillchen) [reply mesop/server/state_session.py:45]: Interestingly, Streamlit has a similar issue: https://github.com/streamlit/streamlit/issues/6602 IIUC, they run into this because their backend is always stateful.

**Review by wwwillchen — APPROVED** [review]
Overall, PR looks good. Main concern here is avoiding out of memory errors. I think it's OK to merge this without enabling by default, and we can ask people to dogfood internally. After we have a more robust story with avoiding out of memory limits, we can flip this as default.

File: mesop/server/config.py
Comment (wwwillchen) [inline mesop/server/config.py:1]: A friendly reminder that we'll need docs for configuring env variable MESOP_STATE_SESSION_BACKEND. Also, Cloud Run session affinity is not enabled by default (https://cloud.google.com/run/docs/configuring/session-affinity) so we'll need to tell people to enable this in our Deployment guide if they use the "memory' option.

**Comment by wwwillchen** [comment]
One more thing: we should run tests in CI with memory and with none - similar to what you did for component tree diffing.

On an adjacent note, I think we could stop testing w/ component tree diffing disabled (to avoid a combinantorial explosion in CI), and probably just remove the enable_component_tree_diffs flag.

File: mesop/server/config.py (hunk 5 above)
Comment (richard-to) [reply mesop/server/config.py]: I agree. That definitely a safer roll out. Updated.

File: mesop/server/state_session.py (hunk 4 above)
Comment (richard-to) [reply mesop/server/state_session.py:45]: Yeah, memory capacity is definitely a concern with python. Even for normal sized objects, Python dicts tend can take up a fair amount of space. Definitely easy for memory to balloon. I think dataclasses are better in that regard. I think with dataclasses if you enable slots, that can also improve memory performance since with slots you can't dynamically add fields to the dataclass which I think we should restrict anyway.

This could be one reason serializing the data into strings even in the memory case could be beneficial. That could reduce the memory load a bit with the trade off of needing to serialize/deserialize each time. But I don't think it would help that much if users are making a lot of interactions per session. But could help with the stale session case.

-----

I like the idea of checking memory. Not ideal but at least we can fall back to sending the request over temporarily, but could lead to uneven performance for users. But better than getting OOM'd.

-----

Ideally, for real use cases, we would have a way to avoid using the memory backend.

File: mesop/server/server.py [hunk 6]
```diff
@@ -241,6 +253,14 @@ def serialize(response: pb.UiResponse) -> str:
   return f"data: {encoded}\n\n"
 
 
+def generate_state_token():
+  """Generates a state token used to cache and look up Mesop state.
+
+  We include a timestamp to reduce the possibility of two identically generated tokens.
+  """
+  return datetime.now().strftime("%Y%m%d%H%M%S%f") + secrets.token_urlsafe(16)
+
```
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/server/server.py:258]: That's true. The datetime could be the same and that's OK. I think the reasoning here is that the combination of the date time + token being the same would have a lower probability of collision than just the token. 

But I agree that the collision chance probably fairly low, so I've removed the date prefix.

File: mesop/protos/ui.proto [hunk 7]
```diff
@@ -127,6 +130,10 @@ message UpdateStateEvent {
         States full_states = 1;
         States diff_states = 2;
     }
+    // If enabled, then we will not send the state back on each request. We will only
+    // send back the state in cases in recovery scenarios.
+    optional bool state_session_enabled = 3;
```
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/protos/ui.proto]: Ah good catch. Yes, this was needed when we had the session id option. But since we removed that, it's not necessary now.

File: mesop/server/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (richard-to) [reply mesop/server/config.py]: Lol. I see. Ok. Yeah. I went with the `type: ignore` approach here.

File: mesop/server/config.py
Comment (richard-to) [reply mesop/server/config.py:1]: Yes, thanks for the reminder. Yes. I plan to do that as a separate change.
